### PR TITLE
feat!: Use strong types for context objects

### PR DESCRIPTION
**Pull request by paulgessinger**
Move away from using a raw typedef to std::any. Doing that has certain drawbacks: std::any can be convert constructed from anything. This causes problems when constructing anything that only accepts a context any. GCC9 even has recursion trouble in the type_traits infrastructure leading to compile time failures.

Moving to wrappers can alleviate this. They only have an explicit constructor, meaning that they are not autoconstructed by conversion. This is a little bit more cumbersome to use, but much more robust I believe.

Performance should not be affected.

Previously, to break the problems of converting constructors with a single `std::any` argument, we resorted to `std::reference_wrapper<std::any>`, which only resolved it to some degree. This PR also removes this workaround. `std::reference_wrapper<const {Geometry,MagneticField}Context>` is still retained when used to store a context reference as a member, as for example in the stepper state. 

BREAKING CHANGE: The public interface of the geometry and magnetic context type change. Before, they were raw `std::any`s, now they are of type `Acts::ContextType`. Construction needs to be explicit, and the contained `std::any` can be accessed via the `Acts::ContextType::get()` methods. `std::any_cast` to unpack the concrete context value is still supported, but needs to retrieve the internal `std::any` first. **NOTE**: If you forget to call `.any()` on the context object, `std::any_cast` will compile, but will throw `std::bad_any_cast` at runtime.

### CODE REVIEW

**Review by HadrienG2 — APPROVED** [review]
LGTM.

File: Core/include/Acts/MagneticField/ConstantBField.hpp
Comment (asalzburger) [inline Core/include/Acts/MagneticField/ConstantBField.hpp:24]: I like, we used this reference wrapping only for the context mechanism, happy when this goes.

Comment (paulgessinger) [reply Core/include/Acts/MagneticField/ConstantBField.hpp:24]: Yeah we still need it when we want to **store** a reference to the context object, which we sometimes do in our state objects, but that's not problematic.

**Review by asalzburger — APPROVED** [review]

**Comment by asalzburger** [comment]
I approve, nice work Paul!

Let's wait for the CI and merge then.

**Comment by paulgessinger** [comment]
I think I'm actually going to implement @benjaminhuth's suggestion before we merge this.

**Comment by paulgessinger** [comment]
Ok, I switched to an access method, and don't expose the `std::any` anymore at all.

File: Examples/Detectors/ContextualDetector/include/ActsExamples/ContextualDetector/AlignedDetectorElement.hpp
Comment (asalzburger) [inline Examples/Detectors/ContextualDetector/include/ActsExamples/ContextualDetector/AlignedDetectorElement.hpp:87]: Yes, looks better.

**Review by asalzburger — APPROVED** [review]
Re-appoving.

**Pull request merged**
